#220 Fix case-sensitivity of HTTP headers in models

parawanderer opened the pull request:
*Issue #, if available:* #220

*Description of changes:*

This solution builds upon @jeromevdl 's solution of PR [234](https://github.com/aws/aws-lambda-java-libs/pull/234) to fix the
issue of headers being case sensitive. Fixes @carlzogh 's comment
regarding breaking changes in setHeaders.

This adds some extra unit tests, and reorders some of the added unit test assertions of the original PR, too


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

msailes commented:
Do you think that `HttpHeadersUtil.mergeOrReplace()` should be a function within the `HttpHeaders` class, instead of within it's own class.

Thoughts @carlzogh @andclt @jeromevdl

parawanderer commented:
> Do you think that `HttpHeadersUtil.mergeOrReplace()` should be a function within the `HttpHeaders` class, instead of within it's own class.
> 
> Thoughts @carlzogh @andclt @jeromevdl

Personally I could agree to that too, since you can just as easily have it be package-private on a method level that way and that was the reason I thought of moving it to a package-private class to begin with

carlzogh commented on aws-lambda-java-events/src/main/java/com/amazonaws/services/lambda/runtime/events/ApplicationLoadBalancerRequestEvent.java:
```diff
@@ -40,8 +41,8 @@
     private Map<String, String> queryStringParameters;
     private Map<String, List<String>> multiValueQueryStringParameters;
-    private Map<String, String> headers;
-    private Map<String, List<String>> multiValueHeaders;
+    private HttpHeaders<String> headers;
+    private HttpHeaders<List<String>> multiValueHeaders;
```
I believe this is a breaking change for some consumers - I suggest we include it as part of the next major version of this library (v4). See https://github.com/aws/aws-lambda-java-libs/issues/282 where changes are being discussed

carlzogh commented on aws-lambda-java-events/src/main/java/com/amazonaws/services/lambda/runtime/events/APIGatewayV2WebSocketResponse.java:
```diff
@@ -12,8 +14,8 @@
-  private Map<String, String> headers;
-  private Map<String, String[]> multiValueHeaders;
+  private HttpHeaders<String> headers;
+  private HttpHeaders<String[]> multiValueHeaders;
```
If we're doing this change as part of v4, we should also change `String[]` to `List<String>` for consistency with other data models

carlzogh commented on aws-lambda-java-tests/src/test/java/com/amazonaws/services/lambda/runtime/tests/EventLoaderTest.java:
```diff
@@ -26,7 +26,6 @@ public void testLoadApiGatewayRestEvent() {
 
         assertThat(event).isNotNull();
         assertThat(event.getBody()).isEqualTo("Hello from Lambda!");
-        assertThat(event.getHeaders()).hasSize(2);
```
Why remove this? Can we not include an equivalent assertion for `HttpHeaders`?

parawanderer replied:
I think I removed this because I created a new test class dedicated solely to testing headers in `aws-lambda-java-tests/src/test/java/com/amazonaws/services/lambda/runtime/tests/HeadersTest.java`, and the header check here did not seem relevant to the name of this set of tests `EventLoaderTest`. At least that's the reasoning that would make sense to me right now.

But since later classes in here do the header test too, right now I'd think there's no real benefit to removing it either. So I'll add it back!

msailes commented:
Hi @parawanderer,

Can this be updated at written as a PR against the [V4](https://github.com/aws/aws-lambda-java-libs/tree/events-v4-serialization-v2) branch?

parawanderer commented:
Hi @msailes

I've taken a look and I believe I merged it in with the "events-v4-serialization-v2" branch, changed things in line with previous suggestions in this thread and updated all relevant tests/usages. 
Please feel free to check this again for any more issues.
